### ISSUE TITLE
Helm search reports inconsistent results with a large index

### ISSUE THREAD

**Issue by sharkymcdongles** [helm/helm#8295]
We use a project called helmsman to manage state in our environments. When it runs it does a helm search of every chart in the desired state file. This search hits often time fails to find charts that exist. 

Our repo is large with 30k+ charts at 28MB in size. When I go to:

`~/.cache/helm/repository/repo-index.yaml`

I can see the chart inside the cache. However it fails. We will try cleaning our repo to help fix this, but there's no documentation or anything saying there's an upper limit on the number of charts a repo can contain, so from my perspective this is a bug.

We have a test case to show this not working properly, please see below:
```
while true; do helm search repo repo/upload-service --version 1.1.0 -o json; done
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"63a7bd5","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"76906aa","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"63a7bd5","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"63a7bd5","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"3b7e831","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
[{"name":"repo/upload-service","version":"1.1.0","app_version":"1.0","description":"A Helm chart for Kubernetes"}]
```

Output of `helm version`:
v3.2.1

Output of `kubectl version`:
v1.16.3

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
GKE

**Comment by mattfarina** [comment]
That's an amazingly large index. I tested with an 8mb index but was unable to produce the problem. Any help is being able to reproduce would help. Also, PRs to fix this are welcome.

**Comment by sharkymcdongles** [comment]
@mattfarina after further investigation it seems the initial suspect of index size was incorrect. It appears that there were some nonsemantic versions uploaded ages ago during the Helm2 times. These 2 charts had versions such as: 880f220 and 76906aa.

Given how the search works with scoring it seems that it could not consistently score the results due to the nonstandard versions leading to it randomly throwing back the incorrect version when searching.

Now given how Helm 3 is stricter about semantic versions, I am still unsure if this is a bug because I don't think you can upload charts with versions such as the ones that caused the issue here. Can you enlighten me or somebody else if it is possible for such charts to be uploaded? If so I could try and work one  MR to fix the way scoring is handled to properly return the correct result even when version is a bit wonky for lack of a better word.

If not, then I suppose we can close this.

**Comment by mattfarina** [comment]
> Now given how Helm 3 is stricter about semantic versions

Is it stricter? Helm v2 required semantic versions. The range syntax changed slightly for edge cases to mirror other semver projects. There is no spec for ranges like there is for the versions.

**Comment by sharkymcdongles** [comment]
I assumed it was stricter because helm 3 introduced the --devel flag. Once
I deleted the malformed versions I listed above it works as expected. So
unsure if this behavior should be addressed and corrected, or if it wasn't
used as expected and is user error. If it IS something that should work I
can see what I can hack up. Just dont want to waste time if it is a mistake
from the end user to even have such versions available in the repo.
